tools/generator/GenerationInfo.cs

```diff
@@ -89,7 +89,9 @@ XElement ToDefineConstants (CodeGeneratorOptions options, XNamespace msbuild)
 		XElement ToCompileElement (XName compile, string path)
 		{
 			path = path.Replace (CSharpDir, "$(MSBuildThisFileDirectory)")
-				.Replace ('/', '\\');
+				.Replace ('/', '\\')
+				.Replace ("$(MSBuildThisFileDirectory)\\", "$(MSBuildThisFileDirectory)");
```

Comment:
This won't properly replace `$(MSBuildThisFileDirectory)\\\Foo`, because it only looks for one `\`, not more than one.

I feel like the better approach is to instead fixup the `GenerationInfo` constructor to ensure that `csdir` *doesn't* end with `Path.DirectorySeparatorChar`:

```csharp
		public GenerationInfo (string csdir, string javadir, string assembly)
		{
			CSharpDir = csdir
				.TrimEnd (Path.DirectorySeparatorChar)
				.TrimEnd (Path.AltDirectorySeparatorChar);
			JavaDir = javadir;
			Assembly = assembly;
		}
```

Reply:
I do not think this scenario can happen.  `path` is created for types via:
```
string filename = Path.Combine (CSharpDir, name + ".cs");			
GeneratedFiles.Add (filename);
```
and for enums via:
```
var path = Path.Combine (output_dir, GetFileName (enu.Key, useShortFileNames) + ".cs");
files.Add (path);
```
I would assume `Path.Combine` is only going to add zero or one slashes.

Reply:
My concern is where `CSharpDir` *comes from*, which is from `generator.exe --csdir=VALUE`.  Thus, if you execute:

```
mono bin/TestDebug/generator.exe --csdir bin/TestDebug/generator-core//// \
    --api-level=20 tools/generator/Tests-Core/api.xml \
    --enummethods=tools/generator/Tests-Core/methods.xml \
    --enumfields=tools/generator/Tests-Core/fields.xml \
    --enumdir=bin/TestDebug/generator-core////
```

then `CSharpDir` will be `bin/TestDebug/generator-core////`.

That said, I'm still wrong, because of the `path.Replace (CSharpDir, "$(MSBuildThisFileDirectory)")` we'd replace `bin/TestDebug/generator-core////` with `$(MSBuildThisFileDirectory)`, thus getting the *desirable* value of e.g. `$(MSBuildThisFileDirectory)Java.Lang.Object.cs` (no intermediate `\`).

We only get the `\` when `generator --cs-dir` *doesn't* end with a directory separator char, due to the `Path.Combine()`, as you noted.